File: 剑指Offer/树的子结构/subTree.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def HasSubtree(self, pRoot1, pRoot2):
        # write code here
        if not pRoot2 or not pRoot1:
            return False
        result = False
        if pRoot1.val == pRoot2.val:
            result =  self.subHasSubtree(pRoot1,pRoot2)
        if not result:
            result = self.HasSubtree(pRoot1.left,pRoot2)
        if not result:
            result = self.HasSubtree(pRoot1.right,pRoot2)
        return result
 
    def subHasSubtree(self,pRoot1,pRoot2):
        if not pRoot2:
            return True
        if not pRoot1:
            return False
        if pRoot1.val != pRoot2.val:
            return False
        if pRoot1.val == pRoot2.val:
            return self.subHasSubtree(pRoot1.left,pRoot2.left) and self.subHasSubtree(pRoot1.right,pRoot2.right)

File: 剑指Offer/树的子结构/subTree.java
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;
 
    public TreeNode(int val) {
        this.val = val;
 
    }
 
}
*/
public class Solution {
    public boolean HasSubtree(TreeNode root1,TreeNode root2) {
        if(root1 == null || root2 == null){
            return false;
        }
        boolean result = false;
        if(root1.val == root2.val){
            result = subHasSubtree(root1,root2);
        }
        if(!result){
            result = subHasSubtree(root1.left,root2);
        }
        if(!result){
            result = subHasSubtree(root1.right,root2);
        }
        return result;
    }
 
    public boolean subHasSubtree(TreeNode root1,TreeNode root2){
        if(root2 == null){
            return true;
        }
        if(root1 == null){
            return false;
        }
        if(root1.val != root2.val){
            return false;
        }
        return subHasSubtree(root1.left,root2.left) && subHasSubtree(root1.right,root2.right);
    }
}